rrdp_iostat.ml

```diff
@@ -313,42 +313,52 @@ module Blktap3_stats_wrapper = struct
 
 	let empty =
 		{
-		st_ds_req       = 0L;
-		st_f_req        = 0L;
-		st_oo_req       = 0L;
-		st_rd_req       = 0L;
-		st_rd_cnt       = 0L;
-		st_rd_sect      = 0L;
-		st_rd_sum_usecs = 0L;
-		st_rd_max_usecs = 0L;
-		st_wr_req       = 0L;
-		st_wr_cnt       = 0L;
-		st_wr_sect      = 0L;
-		st_wr_sum_usecs = 0L;
-		st_wr_max_usecs = 0L;
-		st_mem_mode	= false;
+		read_reqs_submitted     = 0L;
+		read_reqs_completed     = 0L;
+		read_sectors            = 0L;
+		read_total_ticks        = 0L;
+		write_reqs_submitted    = 0L;
+		write_reqs_completed    = 0L;
+		write_sectors           = 0L;
+		write_total_ticks       = 0L;
 		}
 
 	let get_domid_devid_to_stats_blktap3 () : ((int * int) * t) list =
 		let shm_devices_dir = "/dev/shm" in
-		let read_raw_blktap3_stats vbd =
+		let sys_dev_backend_dir = "/sys/bus/xen-backend/devices" in
+		let read_raw_blktap3_stats tdpid domid devid =
 			try
-				let stat_file = Printf.sprintf "%s/%s/statistics" shm_devices_dir vbd in
+				let stat_file = Printf.sprintf "%s/td3-%d/vbd-%d-%d" shm_devices_dir tdpid domid devid in
 				(* Retrieve blktap3 statistics record *)
 				let stat_rec = get_blktap3_stats stat_file in
 				Some stat_rec
 			with _ ->
 				None
 		in
-		let shm_dirs = Array.to_list (Sys.readdir shm_devices_dir) in
+		let shm_dirs = Array.to_list (Sys.readdir sys_dev_backend_dir) in
 		let shm_vbds = List.filter (fun s -> String.startswith "vbd3-" s) shm_dirs in
 		List.fold_left (fun acc vbd ->
+			let domid, devid = Scanf.sscanf vbd "vbd3-%d-%d" (fun id devid -> (id, devid)) in
+			try
+				with_xs (fun xs ->
+					let path = Printf.sprintf "/local/domain/0/backend/vbd3/%d/%d/kthread-pid" domid devid in
+					let tdpid = int_of_string (xs.Xs.read path) in
+					match read_raw_blktap3_stats tdpid domid devid with
+					| Some stat -> ((domid, devid), stat) :: acc
+					| None -> acc
+					)
+			with e ->
+				D.error "Error while looking up tapdisk pid: %s" (Printexc.to_string e);
+				acc
+		) [] shm_vbds
+
+		(* List.fold_left (fun acc vbd ->
 			match read_raw_blktap3_stats vbd with
 			| Some stat ->
 				let domid, devid = Scanf.sscanf vbd "vbd3-%d-%d" (fun id devid -> (id, devid)) in
 				((domid, devid), stat) :: acc
 			| None -> acc
-		) [] shm_vbds
+		) [] shm_vbds *)
```

Comment:
I kept this, since similar logic would be needed when low_mem_mode info is added. Will make the change anyway, thanks